src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -334,117 +350,155 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
 
 		void IDeserializationCallback.OnDeserialization(object sender)
 		{
-			batcher = Factory.Settings.BatcherFactory.CreateBatcher(this, interceptor);
+			_batcher = Factory.Settings.BatcherFactory.CreateBatcher(this, _interceptor);
 		}
 
 		#endregion
 
 		public ITransaction BeginTransaction(IsolationLevel isolationLevel)
 		{
 			Transaction.Begin(isolationLevel);
-			return transaction;
+			return _transaction;
 		}
 
 		public ITransaction BeginTransaction()
 		{
 			Transaction.Begin();
-			return transaction;
+			return _transaction;
 		}
 
 		public ITransaction Transaction
 		{
 			get
 			{
-				if (transaction == null)
+				if (_transaction == null)
 				{
-					transaction = Factory.TransactionFactory.CreateTransaction(session);
+					_transaction = Factory.TransactionFactory.CreateTransaction(Session);
 				}
-				return transaction;
+				return _transaction;
 			}
 		}
 
 		public void AfterNonTransactionalQuery(bool success)
 		{
-			log.Debug("after autocommit");
+			_log.Debug("After autocommit");
 		}
 
 		private bool IsAfterTransactionRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.AfterTransaction; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction;
 
 		private bool IsOnCloseRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.OnClose; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.OnClose;
 
 		private bool IsAggressiveRelease
-		{
-			get
-			{
-				if (connectionReleaseMode == ConnectionReleaseMode.AfterTransaction)
-				{
-					return !IsInActiveTransaction;
-				}
-				return false;
-			}
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction && !IsInActiveTransaction;
 
 		public ISessionFactoryImplementor Factory
-		{
-			get { return session.Factory; }
-		}
+			=> Session.Factory;
 
 		public bool IsReadyForSerialization
 		{
 			get
 			{
-				if (ownConnection)
-				{
-					return connection == null && !batcher.HasOpenResources;
-				}
-				else
+				if (_ownConnection)
 				{
-					return connection == null;
+					return _connection == null && !_batcher.HasOpenResources;
 				}
+				return _connection == null;
 			}
 		}
 
 		/// <summary> The batcher managed by this ConnectionManager. </summary>
 		public IBatcher Batcher
+			=> _batcher;
+
+		public DbCommand CreateCommand()
 		{
-			get { return batcher; }
+			var result = GetConnection().CreateCommand();
+			Transaction.Enlist(result);
+			return result;
 		}
 
-		public IDisposable FlushingFromDtcTransaction
+		/// <summary>
+		/// Enlist the connection into provided transaction if the connection should be enlisted.
+		/// Does nothing in case an explicit transaction ongoing.
+		/// </summary>
+		/// <param name="transaction">The transaction in which the connection should be enlisted.</param>
+		public void EnlistIfRequired(System.Transactions.Transaction transaction)
+		{
+			if (_connection == null ||
+				_connectionReleaseRequired ||
```

Comment:
If slated for releasing, do not enlist it, a new connection will be acquired at next usage. Enlisting a "to be released connection" can cause transaction cancellation with SQL-Server, because such a connection may still be considered enlisted with previous transaction, and this somewhat "contaminates" the new scope. (Gives a "System.Data.SqlClient.SqlException : Microsoft Distributed Transaction Coordinator (MS DTC) has stopped this transaction." in distributed tests.)